numpy/core/src/multiarray/item_selection.c

```diff
@@ -1749,6 +1750,281 @@ PyArray_LexSort(PyObject *sort_keys, int axis)
     return NULL;
 }
 
+/*
+* Helper function for PyArray_KeySort
+* Sorts 1D slice based on argsort indices
+* takes in raw bytes of data, indices array
+* Number of elements, and stride of data
+* element size and temporary storage for an element.
+*/
+void reorder_by_index(char *data, char *current, npy_intp *index, npy_intp N, npy_intp stride, npy_intp elsize) {
+    // Reorder the indices and relevant values cyclically
```

Comment:
Not sure if it helps without thinking too much about it, but maybe we can write: The following code reorders the data with respect to the index. The inner `while` loop places a single element to the right place until it reaches a fully cycle (an already ordered element). Each element may be visited twice, but will be sorted on the first visit.  The second visit finds it already sorted and immediately continues.